### PR TITLE
TxSummary digest

### PR DESCRIPTION
**Pull request by garbageslam**
A proposal to introduce a `TxSummary` object and incorporate it into the digest which MLSAG's sign, to help hardware wallets.

[Rendered Proposal](https://github.com/garbageslam/mcips/blob/tx_summary_digest/text/0052-tx-summary-digest.md)

### CODE REVIEW

File: text/0052-tx-summary-digest.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+A `TxSummary` is an object much like the `TxPrefix`, but where instead of the `TxIn`
+objects, the corresponding `pseudo_output_commitments` are stored.
+
+There is exactly one possible `TxSummary` for a given `Tx`.
+
+The `TxSummary` contains:
+* The list of outputs (omitting "extra" parts like encrypted fog hint and encrypted memo).
+* The list of pseudo_output_commitments (commitments to the values of the true inputs)
+  * Any input rules associated to these inputs (if they are [MCIP 42](https://github.com/mobilecoinfoundation/mcips/pull/42) signed contingent inputs)
```
Comment (eranrund) [inline text/0052-tx-summary-digest.md:59]: Isn't there a concern that these rules might also be large, since they contain complete `TxOut`s?

**Review by eranrund — COMMENTED** [review]
This looks reasonable to me, thank you for the detailed write up.
Do you think it would make sense to add a "worst possible case" calculation of how much memory the hardware wallet will need to support, based on a transaction that has the max inputs/outputs?

File: text/0052-tx-summary-digest.md (hunk 1 above)
Comment (garbageslam) [reply text/0052-tx-summary-digest.md:59]: yeah probably we don't need all the rules -- maybe we just need like, the indices of the TxOut's that actually were relevant to this SCI

good point, thanks

**Comment by garbageslam** [comment]
> Do you think it would make sense to add a "worst possible case" calculation of how much memory the hardware wallet will need to support, based on a transaction that has the max inputs/outputs?

Koe made a different estimate of the sizes of Tx's due to membership proofs which is larger than mine, in MCIP #21 

> Currently (protocol v0), Merkle membership proofs constitute the bulk of transaction size. A 1-input/1-output transaction is \~21 kB, with membership proofs taking up \~17 kB (assuming membership proofs with 32 elements, representing a Merkle tree 31 layers deep). In a 16-input/16-output transaction (the maximum size of a transaction), membership proofs occupy \~275 kB out of \~325 kB.

I have not actually done this calculation myself.

We could also write a unit test or something that measures this maximum size.

You're right, we should probably document what the actual worst possible case is at the current version (v2)

**Comment by eranrund** [comment]
> > Do you think it would make sense to add a "worst possible case" calculation of how much memory the hardware wallet will need to support, based on a transaction that has the max inputs/outputs?
> 
> Koe made a different estimate of the sizes of Tx's due to membership proofs which is larger than mine, in MCIP #21
> 
> > Currently (protocol v0), Merkle membership proofs constitute the bulk of transaction size. A 1-input/1-output transaction is ~21 kB, with membership proofs taking up ~17 kB (assuming membership proofs with 32 elements, representing a Merkle tree 31 layers deep). In a 16-input/16-output transaction (the maximum size of a transaction), membership proofs occupy ~275 kB out of ~325 kB.
> 
> I have not actually done this calculation myself.
> 
> We could also write a unit test or something that measures this maximum size.
> 
> You're right, we should probably document what the actual worst possible case is at the current version (v2)

I think it's also valuable (if not more valuable) to document how much it would be following this change. This will make it easier for potential hardware integrations to have a quick go/no-go based on memory requirements.

File: text/0052-tx-summary-digest.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+to what addresses as a result of this transaction, and that the device can verify
+that this information is correct.
+
+However, with the status quo, the only way that this can be accomplished is to put
+the entire `Tx` on the device so that it can verify the `extended_message_digest`,
+because the `extended_message_digest` is the only thing that connects the `MLSAG`
+that the device is signing to the rest of the `Tx`.
+
+This increases the amount of data that has to be sent to the device from a few kb
+to `> 100kb` in the worst case. This is because Merkle proofs are each e.g. 40 bytes
```
Comment (remoun) [inline text/0052-tx-summary-digest.md:32]: Nit: Capitalize KB
```suggestion
This increases the amount of data that has to be sent to the device from a few KB
to >100 KB in the worst case. This is because Merkle proofs are each ~40 bytes
```

File: text/0052-tx-summary-digest.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+A `TxSummary` is an object much like the `TxPrefix`, but where instead of the `TxIn`
+objects, the corresponding `pseudo_output_commitments` are stored.
+
+There is exactly one possible `TxSummary` for a given `Tx`.
+
+The `TxSummary` contains:
+* The list of outputs (omitting "extra" parts like encrypted fog hint and encrypted memo).
+* The list of pseudo_output_commitments (commitments to the values of the true inputs)
+  * Any input rules associated to these inputs (if they are [MCIP 31](https://github.com/mobilecoinfoundation/mcips/pull/31) signed contingent inputs)
```
Comment (remoun) [inline text/0052-tx-summary-digest.md:59]: Optional/Equivalent: Link to the doc rather than PR
```suggestion
  * Any input rules associated to these inputs (if they are [MCIP 31](0031-transactions-with-contingent-inputs.md) signed contingent inputs)
```

File: text/0052-tx-summary-digest.md [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+TxOutSummary: MaskedAmount masked_amount
+TxOutSummary: from(&TxOut)
+TxInSummary: CompressedCommitment pseudo_output_commitment
+TxInSummary: Option~InputRules~ input_rules
+```
+
+A hardware wallet which is asked to sign an MLSAG can expect to see that 32 byte digest
+and the `TxSummary`.
+
+**Security**:
```
Comment (remoun) [inline text/0052-tx-summary-digest.md:167]: This merits a header IMO
```suggestion
## Security
```

File: text/0052-tx-summary-digest.md [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+digested before that. Then we would not need to have an extra round of hashing things at the end.
+
+However, we do not propose at this time to completely change the `Tx` structure to fix these issues, as this would be
+a very complicated breaking change.
+
+Instead, we view this current proposal as the simplest and most straightforward proposal that will meet the needs of
+hardware wallets and help them to avoid complexities around streaming too much data and having to create a state machine to
+compute a proper merlin digest of the entire `Tx` object, which would also make the user experience significantly slower.
+
+## SCI support, definition of `TxSummaryUnblindingData`.
```
Comment (remoun) [inline text/0052-tx-summary-digest.md:227]: Nit: The other headings do not end in a period
```suggestion
## SCI support, definition of `TxSummaryUnblindingData`
```

File: text/0052-tx-summary-digest.md [hunk 6]
```diff
@@ -0,0 +1,253 @@
+- Feature Name: tx_summary_digest
+- Start Date: 2022-10-08
+- MCIP PR: [mobilecoinfoundation/mcips#0052](https://github.com/mobilecoinfoundation/mcips/pull/0052)
+- Tracking Issue: [mobilecoinfoundation/mobilecoin#2686](https://github.com/mobilecoinfoundation/mobilecoin/issues/2686)
+
+# Summary
+[summary]: #summary
+
+Introduce an object called `TxSummary` which is significantly smaller than a `Tx`,
```
Comment (remoun) [inline text/0052-tx-summary-digest.md:9]: Nit: "an object called" doesn't add much
```suggestion
Introduce a `TxSummary` that is significantly smaller than a `Tx`,
```

**Review by remoun — APPROVED** [review]
LGTM

File: text/0052-tx-summary-digest.md [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+**Security**:
+From a security point of view, the hardware wallet can know that it is intractable
+for someone to find a different `TxSummary` that produces the same `extended_message_and_tx_summary` digest,
+so if it signs the MLSAG, this is the only thing that it can be committing to.
+The hardware wallet also knows how the validators will compute the `extended_message_and_tx_summary` digest
+based on the `Tx`, and knows that it is infeasible for anyone to find another `Tx` that has the same digest here.
+So, an attacker (on the computer) could give the device an improperly formed `TxSummary` and lie to the device
+this way, but the attacker will not be able to get consensus to accept those signatures. The device therefore
+knows that the either the `TxSummary` is accurate, or it's signature doesn't matter because the `Tx` will not
```
Comment (nick-mobilecoin) [inline text/0052-tx-summary-digest.md:175]: :upside_down_face: 
```suggestion
knows that either the `TxSummary` is accurate, or it's signature doesn't matter because the `Tx` will not
```

**Review by nick-mobilecoin — APPROVED** [review]

File: text/0052-tx-summary-digest.md [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+However, we do not propose at this time to completely change the `Tx` structure to fix these issues, as this would be
+a very complicated breaking change.
+
+Instead, we view this current proposal as the simplest and most straightforward proposal that will meet the needs of
+hardware wallets and help them to avoid complexities around streaming too much data and having to create a state machine to
+compute a proper merlin digest of the entire `Tx` object, which would also make the user experience significantly slower.
+
+## SCI support, definition of `TxSummaryUnblindingData`.
+
+We have tried to future-proof this design against [MCIP 31](https://github.com/mobilecoinfoundation/mcips/pull/31) Signed Contingent Inputs being present in the `Tx`.
```
Comment (isis-mc) [inline text/0052-tx-summary-digest.md:229]:
```suggestion
We have tried to future-proof this design against [MCIP 31](0031-transactions-with-contingent-inputs.md) Signed Contingent Inputs being present in the `Tx`.
```

**Review by isis-mc — APPROVED** [review]
LGTM

File: text/0052-tx-summary-digest.md [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+* The tombstone block
+
+During transaction construction, the signer computes the extended message digest as before,
+but now creates a merlin transcript using that 32-byte digest, followed by digesting the
+`TxSummary`. The MLSAG's (except for SCIs) will sign 32-bytes extracted from this digest.
+The verifier similarly computes this
+`extended-message-and-tx-summary digest` and verifies that the MLSAGs sign this.
+
+When a hardware wallet is asked to sign an MLSAG, we can give it now the `extended_message_digest`
+and the `TxSummary`, and it can compute the appropriate digest from this for the MLSAG to sign.
```
[the remark this answers is not in the document]
Comment (garbageslam) [reply text/0052-tx-summary-digest.md]: does the rust merlin implementation not work? can we just fix that one?

File: text/0052-tx-summary-digest.md [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+TxSummary: uint64 fee_value
+TxSummary: uint64 fee_token_id
+TxSummary: uint64 tombstone_block
+TxSummary: new(&TxPrefix, &[CompressedCommitment])
+TxOutSummary: CompressedRistrettoPublic public_key
+TxOutSummary: CompressedRistrettoPublic target_key
+TxOutSummary: MaskedAmount masked_amount
+TxOutSummary: bool associated_to_input_rules
+TxInSummary: CompressedCommitment pseudo_output_commitment
+TxInSummary: bool has_input_rules
```
Comment (garbageslam) [inline text/0052-tx-summary-digest.md:181]: Ryan made a suggestion that, maybe the `TxInSummary` should include a digest of the input rules rather than just this flag. Then if later, we think it's important to pass all the rules, we can do so without an enclave change. So it's a bit more future proof, and the cost in terms of wire protocol seems very low.

File: text/0052-tx-summary-digest.md [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+* It can be streamed to a small hardware device, together with the `extended_message_digest`,
+  in order to efficiently compute the (new) digest which MLSAG's must sign.
+* Small additional data (`TxSummaryUnblindingData`) can be streamed with the pieces of the `TxSummary`, allowing
+  the device to "unblind" the `TxSummary`.
+* This can be used to create a `TxSummaryUnblindingReport`, which deduces the following:
+  * For each output, the amount, and the entity to which it is destined, which is one of three possibilities:
+    * The signer themselves
+    * A known contact, identified by their public address
+    * An anonymous swap counterparty
+  * For each input, the amount, and the entity from whcih it came, which is one of two possibilities:
```
Comment (eranrund) [inline text/0052-tx-summary-digest.md:65]:
```suggestion
  * For each input, the amount, and the entity from which it came, which is one of two possibilities:
```

File: text/0052-tx-summary-digest.md [hunk 12]
```diff
@@ -0,0 +1,510 @@
...
+`extended-message-and-tx-summary digest` and verifies that the MLSAGs sign this.
+
+**Note**: This means that the `TxSummary` needs to be constructible by the consensus enclave, given only
+the `Tx`, and therefore cannot contain any of the transaction secrets needed to unblind the `TxSummary`. These
+secrets appear in `TxSummaryUnblindingData` instead. The host computer (ultimately the `TransactionBuilder`)
+is responsible to compute this when it creates
+a `Tx`, and then it can be streamed to, and verified by, the hardware device which is asked to sign.
+
+When a hardware wallet is asked to sign an MLSAG, we can give it now the `extended_message_digest`
+and steam it the `TxSummary` and the `TxSummaryUnblindingData`, and it can compute the appropriate digest from this for the MLSAG to sign.
```
Comment (eranrund) [inline text/0052-tx-summary-digest.md:101]:
```suggestion
and stream it the `TxSummary` and the `TxSummaryUnblindingData`, and it can compute the appropriate digest from this for the MLSAG to sign.
```

File: text/0052-tx-summary-digest.md [hunk 13]
```diff
@@ -0,0 +1,510 @@
...
+  If all of these are a match for the `TxOutSummary`, then the device has established that the `TxOut` was originally addressed to
+  this person with this amount -- it is discrete-log hard to find another amount with the same commitment. It is also generally
+  impossible to make a TxOut that belongs to two addresses -- starting from two addresses with different view public keys,
+  finding two different tx_private keys that produce the same tx out public key implies knowledge of a linear relation between
+  the view private keys. Also obtaining a collision of tx target keys requires finding a hash collision which has negligible success
+  probability.
+* In the final case, the `TxOutUnblindingData` doesn't contain an address or `tx_private_key`. This can occur normally when the `TxOut`
+  was not created using `TransactionBuilder.add_output` interface, but rather
+  came as a required (or partial fill) output from a Signed Contingent Input which was added to the `Tx`.
+  In this case it is not addressed to ourself, and due to the anonyminity properties of the MCIP 31 atomic swaps, we cannot know who
```
Comment (eranrund) [inline text/0052-tx-summary-digest.md:257]:
```suggestion
  In this case it is not addressed to ourself, and due to the anonymity properties of the MCIP 31 atomic swaps, we cannot know who
```

File: text/0052-tx-summary-digest.md [hunk 14]
```diff
@@ -0,0 +1,510 @@
...
+  We can compute the corresponding Pedersen commitment to these deterministically, and check if this matches
+  the pseudo-output commitment in the `TxInSummary`. This came from the `TxSummary` which is checked by the enclave,
+  and so we know that it must be accurate. If we manage to unblind it, then we know this is the correct unblinding,
+  because it is discrete-log hard to find another unmasking of the Pedersen commitment. If we can't unblind it, then
+  the hardware device knows that the computer has lied / provided a false unmasking, but there is no legitimate reason
+  for the computer to fail to do this, because regardless of whether this is our own input or a signed input from a
+  counterparty, we can expect to have the unmasked amount or there is no way we can possibly spend it.
+* Now having the amount in hand, the next step is to check the `has_input_rules` flag from `TxInSummary`. We know that
+  this flag is accurate because the consensus enclave checks the `TxSummary` and can see which inputs actually have rules.
+  If there are no rules, then we know that we must be signing for this transaction. If there are rules then we attribute
```
Comment (eranrund) [inline text/0052-tx-summary-digest.md:280]: signing for this transaction or for this input?

File: text/0052-tx-summary-digest.md [hunk 15]
```diff
@@ -0,0 +1,510 @@
...
+
+The following analysis explains why these attempts fail:
+* For `TxOut`'s, the decision tree is specified so that we view-key match it first, and if it matches to use, we short circuit and declare it to be
+  a credit to ourself, regardless of whether it is associated to input rules.
+* For inputs, simply attaching input rules to an input, and then later asking the device to sign it as if there are no input rules, will not work.
+  In MCIP 31, inputs that have associated input rules have to sign a different digest that does not depend on the entire Tx. If instead the entire Tx
+  digest is signed, then the transaction will fail validation.
+
+There is an additional possibility: Suppose that the hardware wallet has features that also support signing the MCIP 31 SCIs. Then it could be possible
+that we authorize an sign an SCI, and then later we build a transaction that matches against it, so that we are essentially "trading with ourselves".
```
Comment (eranrund) [inline text/0052-tx-summary-digest.md:309]:
```suggestion
that we authorize and sign an SCI, and then later we build a transaction that matches against it, so that we are essentially "trading with ourselves".
```

**Review by eranrund — APPROVED** [review]
This is really well written. Thank you!

File: text/0052-tx-summary-digest.md (hunk 14 above)
Comment (garbageslam) [reply text/0052-tx-summary-digest.md:280]: thanks

**Comment by garbageslam** [comment]
Going to hold this one an extra day due to comments
